lib/Conversion/SeqToSV/FirRegLowering.h

```diff
@@ -10,26 +10,127 @@
 #ifndef CONVERSION_SEQTOSV_FIRREGLOWERING_H
 #define CONVERSION_SEQTOSV_FIRREGLOWERING_H
 
+#include "circt/Dialect/Comb/CombOps.h"
 #include "circt/Dialect/HW/HWOps.h"
 #include "circt/Dialect/SV/SVOps.h"
 #include "circt/Dialect/Seq/SeqOps.h"
 #include "circt/Support/LLVM.h"
 #include "circt/Support/Namespace.h"
 #include "circt/Support/SymCache.h"
+#include "mlir/IR/Visitors.h"
+#include <mlir/IR/ValueRange.h>
+#include <stack>
+#include <unordered_set>
 
 namespace circt {
+
+using namespace hw;
+// This class computes the set of muxes that are reachable from an op.
+// The heuristic propagates the reachability only through the 3 ops, mux,
+// array_create and array_get. All other ops block the reachability.
+// This analysis is built lazily on every query.
+// The query: is a mux is reachable from a reg, results in a DFS traversal
+// of the IR rooted at the register. This traversal is completed and the result
+// is cached in a Map, for faster retrieval on any future query of any op in
+// this subgraph.
+class ReachableMuxes {
+public:
+  ReachableMuxes(HWModuleOp m) : module(m) {}
+
+  bool isMuxReachableFrom(Operation *regOp, Operation *muxOp) {
```

Comment:
minor, but maybe instead of passing `Operation *`, pass the actual operation subclass, so we can call methods on it directly instead of things like `regOp->getResult(0)`.